Packaging guide rewrite (#50884)
LLNL/spack: lib/spack/docs/packaging_guide_1.rst

#. **Specs**: expressions for describing builds of software, and
#. **Packages**: Python modules that describe how to build and test software according to a spec.

Specs allow a user to describe a *particular* build in a way that a package author can understand.
Packages allow the packager to encapsulate the build logic for different versions, compilers, options, platforms, and dependency combinations in one place.
Essentially, a package translates a spec into build logic.
It also allows the packager to write spec-specific tests of the installed software.

Packages in Spack are written in pure Python, so you can do anything in Spack that you can do in Python.
Python was chosen as the implementation language for two reasons. 
First, Python is ubiquitous in the scientific software community.
Second, it has many powerful features to help make package writing easy.

.. _setting-up-for-package-development:

----------------------------------
Setting up for package development
----------------------------------

For developing new packages or working with existing ones, it's often helpful to have the ``spack/spack-packages`` repository in a custom location.

> By default, Spack will download the builtin package repository to ``~/.spack/package_repos/<hash>/``, which can be inconvenient for development purposes.

Reviewer: From a usability (and support) standpoint, we should discourage package contributors from relying on the default location. Instead encouraging them to clone a copy of the repository in a different location.

> The best approach is to configure Spack to use a custom repository location, such as ``~/spack-packages``:

Reviewer: ```suggestion
We recommend working from a fork of the packages repository.  (See `GitHub <https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/fork-a-repo>`_.) You will need to configure Spack to use it. For example, suppose you have a fork on GitHub called ``YourGHAccount/spack-packages`` and you want to work off of your home directory. Further suppose you want to call the remote Git repository ``upstream``. You would set up your clone and configure Spack as follows:
```

.. code-block:: console

>    $ spack repo set --destination ~/spack-packages builtin

Reviewer: ```suggestion
   $ cd ~
   $ git clone git@github.com:YourGHAccount/spack-packages.git
   $ cd spack-packages
   $ git remote add --track develop upstream git@github.com:spack/spack-packages.git
   $ spack repo set --destination ~/spack-packages builtin
```

Author: Rephrased the whole section. Personally I think Spack shouldn't turn into a git / github tutorial, maybe a few commands is fine though. The `--track develop` bit is useful yeah.

Reviewer: I only added that because, in interacting with users, I have had to point people to the documentation and or provided the commands.

You'll notice that OpenMPI's ``url_for_version()`` function makes use of a special ``Version`` function called ``up_to()``.
When you call ``version.up_to(2)`` on a version like ``1.10.0``, it returns ``1.10``. ``version.up_to(1)`` would return ``1``.
This can be very useful for packages that place all ``X.Y.*`` versions in a single directory and then places all ``X.Y.Z`` versions in a sub-directory.

There are a few ``Version`` properties you should be aware of.
We generally prefer numeric versions to be separated by dots for uniformity, but not all tarballs are named that way.
For example, ``icu4c`` separates its major and minor versions with underscores, like ``icu4c-57_1-src.tgz``.
The value ``57_1`` can be obtained with the use of the ``version.underscored`` property.
There are other separator properties as well:

===================  ======
Property             Result
===================  ======
version.dotted       1.2.3
version.dashed       1-2-3
version.underscored  1_2_3
version.joined       123
===================  ======

.. note::

   Python properties don't need parentheses. ``version.dashed`` is correct.

>    ``version.dashed()`` is incorrect.

Reviewer 2: I think this note is redundant:
```suggestion
```

Author: Feel free to push changes directly.

Spack can create the boilerplate for new packages and open them in your editor for you to fill in.
It can also help you edit existing packages, so you don't have to navigate to the package directory manually.

.. _controlling-the-editor:

^^^^^^^^^^^^^^^^^^^^^^
Controlling the editor
^^^^^^^^^^^^^^^^^^^^^^

When Spack needs to open an editor for you (e.g., for commands like :ref:`cmd-spack-create` or :ref:`cmd-spack-edit`), it looks at several environment variables to figure out what to use.
The order of precedence is:

* ``SPACK_EDITOR``: highest precedence, in case you want something specific for Spack;
* ``VISUAL``: standard environment variable for full-screen editors like ``vim`` or ``emacs``;
* ``EDITOR``: older environment variable for your editor.

You can set any of these to the command you want to run, e.g., in ``bash`` you might run one of these::

  export VISUAL=vim
  export EDITOR="emacs -nw"
  export SPACK_EDITOR=nano

If Spack finds none of these variables set, it will look for ``vim``, ``vi``, ``emacs``, ``nano``, and ``notepad``, in that order.

.. _cmd-spack-create:

^^^^^^^^^^^^^^^^^^^^^
Creating new packages
^^^^^^^^^^^^^^^^^^^^^

> To create a new package Spack provides a command that generates a ``package.py`` file in an existing repository, with a boilerplate package template.

Reviewer 3: ```suggestion
To create a new package, Spack provides a command that generates a ``package.py`` file in an existing repository, with a boilerplate package template.
```

For example, ``icu4c`` separates its major and minor versions with underscores, like ``icu4c-57_1-src.tgz``.
The value ``57_1`` can be obtained with the use of the ``version.underscored`` property.
There are other separator properties as well:

===================  ======
Property             Result
===================  ======
version.dotted       1.2.3
version.dashed       1-2-3
version.underscored  1_2_3
version.joined       123
===================  ======

.. note::

   Python properties don't need parentheses. ``version.dashed`` is correct.
   ``version.dashed()`` is incorrect.

In addition, these version properties can be combined with ``up_to()``.
For example:

.. code-block:: python

   >>> version = Version("1.2.3")
   >>> version.up_to(2).dashed
   Version("1-2")
   >>> version.underscored.up_to(2)
   Version("1_2")

As you can see, order is not important.
Just keep in mind that ``up_to()`` and the other version properties return ``Version`` objects, not strings.

If a URL cannot be derived systematically, or there is a special URL for one of its versions, you can add an explicit URL for a particular version:

.. code-block:: python

>    version("8.2.1", md5="4136d7b4c04df68b686570afa26988ac", url="http://example.com/foo-8.2.1-special-version.tar.gz")

Reviewer 3: This should show a sha256 instead, since we don't accept md5 sums anymore.

When you supply a custom URL for a version, Spack uses that URL *verbatim* and does not perform extrapolation.
The order of precedence of these methods is:

#. package-level ``url``
#. ``url_for_version()``
#. version-specific ``url``

so if your package contains a ``url_for_version()``, it can be overridden by a version-specific ``url``.

If your package does not contain a package-level ``url`` or ``url_for_version()``, Spack can determine which URL to download from even if only some of the versions specify their own ``url``.
Spack will use the nearest URL *before* the requested version.
This is useful for packages that have an easy to extrapolate URL, but keep changing their URL format every few releases.

> With this method, you only need to specify the ``url`` when the URL changes.

Reviewer 3: I think we should add an example of this usage to make it clearer how it works

Author: I don't know if this is how it really works :p, might check that later... But now it feels like enough changes were made.

^^^^^^^^^^^^^^^^^^^^^
Checksum verification
^^^^^^^^^^^^^^^^^^^^^

In the above example we see that each version is associated with a ``sha256`` checksum.
Spack verifies that the source code it downloads is not corrupted or compromised since it was packaged, by verifying that the checksum corresponds to what the original author of the Spack package saw when they created the package.

If Spack uses a download method it can verify, we say the download method is *trusted*. 
Trust is important for *all downloads*: Spack has no control over the security of the various sites from which it downloads source code, and can never assume that any particular site hasn't been compromised.

Trust is established in different ways for different download methods.
For the most common download method --- a single-file tarball --- the tarball is checksummed.

> Git downloads using ``commit=`` are trusted implicitly, as long as a full commit sha is specified.

Reviewer 3: We haven't yet discussed git-based versions at all here -- we should probably move that to before this section or provide a link here to where we cover it later.

Leaning somewhat towards the latter for flow reasons, could be something like:

```suggestion
Later we will discuss :ref:`git-based version declarations <git-fetch>`. Git versions that specify a full commit sha are trusted implicitly, because ``git`` handles checksum verification itself.
```

It could also be worth adding a note to the end of the "Source code and versions" section with something like:

```
Spack supports downloading source code from both :ref:`urls <versions-and-fetching>` and :ref:`version control systems <vcs-fetch>`. 
All the features below that are not specifically related to downloading source code work with either type of version declaration.
```

because we cover a lot of general version features before we circle back to cover versions from VCS systems.

   If that's not possible, e.g., if the package's developers don't name their source archive consistently, you'll need to manually add ``version`` calls yourself.

By default, Spack will search for new versions by scraping the parent URL component of the source archive you gave it in the ``url`` attribute.
So, if the sources are at ``http://example.com/downloads/foo-1.0.tar.gz``, Spack computes a *list URL* from it ``http://example.com/downloads/``, and scans that for links to other versions of the package.
If you need to search another path for download links, you can supply some extra attributes that control how your package finds new versions.
See the documentation on :ref:`attribute_list_url` and :ref:`attribute_list_depth`.

.. _attribute_list_url:

""""""""""""
``list_url``
""""""""""""

This optional attribute can be set to tell Spack where to scan for links to other versions of the package.
For example, ``libdwarf`` has the homepage as the ``list_url``, because that is where links to old versions are:

.. code-block:: python
   :linenos:

>    class Libdwarf(Package):

Reviewer 3: This is a somewhat confusing example, since the parent dir to the homepage is the same as the parent dir to the url.

instead of just a single ``url``.
This attribute is a list of possible URLs that will be tried in order when fetching packages.
Notice that either one of ``url`` or ``urls`` can be present in a package, but not both at the same time.

A well-known case of packages that can be fetched from multiple mirrors is that of GNU.
For that, Spack goes a step further and defines a mixin class that takes care of all of the plumbing and requires packagers to just define a proper ``gnu_mirror_path`` attribute:

.. literalinclude:: .spack/spack-packages/repos/spack_repo/builtin/packages/autoconf/package.py
   :lines: 9-18

^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Preferring versions over others
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

When users install a package without constraining the versions, Spack will typically pick the latest version available.
Usually this is the desired behavior, but as a packager you may know that the latest version is not mature enough or has known issues that make it unsuitable for production use.
In this case, you can mark an older version as preferred using the ``preferred=True`` argument in the ``version`` directive, so that Spack will default to the latest *preferred* version.

> 

Reviewer 3: ```suggestion
See the section on :ref:`version ordering <version-comparison>` for more details and exceptions on how the latest version is computed.

```

.. code-block:: python

   class Foo(Package):
       version("2.0.0", sha256="...")
       version("1.2.3", sha256="...", preferred=True)

.. _deprecate:

^^^^^^^^^^^^^^^^^^^^^^^^
Deprecating old versions
^^^^^^^^^^^^^^^^^^^^^^^^

There are many reasons to remove old versions of software:

#. Security vulnerabilities (most serious reason)
#. Changing build systems that increase package complexity
#. Changing dependencies/patches/resources/flags that increase package complexity
#. Maintainer/developer inability/unwillingness to support old versions
#. No longer available for download (right to be forgotten)

> #. Package or version rename

Reviewer 3: ```suggestion
#. Security vulnerabilities (most serious reason)
#. No longer available for download (right to be forgotten)
#. Maintainer/developer inability/unwillingness to support old versions
#. Changing build systems that increase package complexity
#. Changing dependencies/patches/resources/flags that increase package complexity
#. Package or version rename
```
Trying to reorder to prioritize the list by seriousness, do you agree with this order?

At the same time, there are many reasons to keep old versions of software:

#. Reproducibility
#. Requirements for older packages (e.g., some packages still rely on Qt 3)

In general, you should not remove old versions from a ``package.py`` directly.
Instead, you should first deprecate them using the following syntax:

.. code-block:: python

   version("1.2.3", sha256="...", deprecated=True)

This has two effects.
First, ``spack info`` will no longer advertise that version.
Second, commands like ``spack install`` that fetch the package will require user approval:

.. code-block:: console

   $ spack install openssl@1.0.1e
   ==> Warning: openssl@1.0.1e is deprecated and may be removed in a future Spack release.
   ==>   Fetch anyway? [y/N]

If you use ``spack install --deprecated``, this check can be skipped.

This also applies to package recipes that are renamed or removed.
You should first deprecate all versions before removing a package. If you need to rename it, you can deprecate the old package and create a new package at the same time.

> Version deprecations should always last at least one release of the builtin package repository cycle before the version is completely removed.

Reviewer 3: ```suggestion
Version deprecations should always last at least one release cycle of the builtin package repository before the version is completely removed.
```

``1.2 < 1.2.1 < 1.2.2 < 1.10``.

Other separators for components are also possible, for example ``2025-03-01 < 2025-06``.

Spack can also support string components such as ``1.1.1a`` and ``1.y.0``.
String components are considered less than numeric components, so ``1.y.0 < 1.0``.
This is for consistency with `RPM <https://bugzilla.redhat.com/show_bug.cgi?id=50977>`_.
String components do not have to be separated by dots or any other delimiter.
So, the contrived version ``1y0`` is identical to ``1.y.0``.

Pre-release suffixes also contain string parts, but they are handled in a special way.
For example ``1.2.3alpha1`` is parsed as a pre-release of the version ``1.2.3``.
This allows Spack to order it before the actual release: ``1.2.3alpha1 < 1.2.3``.
Spack supports alpha, beta and release candidate suffixes: ``1.2alpha1 < 1.2beta1 < 1.2rc1 < 1.2``.
Any suffix not recognized as a pre-release is treated as an ordinary string component, so ``1.2 < 1.2-mysuffix``.

Finally, there are a few special string components that are considered "infinity versions".
They include ``develop``, ``main``, ``master``, ``head``, ``trunk``, and ``stable``.

> For example: ``1.2 < develop``.

Reviewer 3: ```suggestion
They include ``develop``, ``main``, ``master``, ``head``, ``trunk``, and ``stable``, in descending order.
For example: ``1.2 < main < develop``.
```

These are useful for specifying the most recent development version of a package (often a moving target like a git branch), without assigning a specific version number.
Infinity versions are not automatically used when determining the latest version of a package unless explicitly required by another package or user.

More formally, the order on versions is defined as follows.

> A version string is split into a list of components based on delimiters such as ``.`` and ``-`` and string boundaries.

Reviewer 3: ```suggestion
A version string is split into a list of components based on delimiters such as ``.``, ``-``, ``_``, and string boundaries.
```

To fetch a package from a source repository, Spack needs to know which VCS to use and where to download from. Much like with ``url``, package authors can specify a class-level ``git``, ``hg``, ``svn``, or ``cvs`` attribute containing the correct download location.

Many packages developed with Git have both a Git repository as well as release tarballs available for download. Packages can define both a class-level tarball URL and VCS.
For example:

.. code-block:: python

   class Trilinos(CMakePackage):

       homepage = "https://trilinos.org/"
       url      = "https://github.com/trilinos/Trilinos/archive/trilinos-release-12-12-1.tar.gz"
       git      = "https://github.com/trilinos/Trilinos.git"

       version("develop", branch="develop")
       version("master",  branch="master")
       version("12.12.1", md5="ecd4606fa332212433c98bf950a69cc7")
       version("12.10.1", md5="667333dbd7c0f031d47d7c5511fd0810")
       version("12.8.1",  md5="9f37f683ee2b427b5540db8a20ed6b15")

If a package contains both a ``url`` and ``git`` class-level attribute,
Spack decides which to use based on the arguments to the ``version()``

> directive. Versions containing a specific branch, tag, or revision are

Reviewer 3: ```suggestion
directive. Versions containing a specific branch, tag, commit, or revision are
```

recommend that all ``tag`` entries be paired with ``commit``. Providing the full 
``commit`` SHA hash allows for Spack to preserve binary provenance for all binaries. 
This is due to the fact that git tags and branches are mutable references to commits, 
but git commits are guaranteed to be unique points in the git history.

The destination directory for the clone is the standard stage source path.

Default branch
  To fetch a repository's default branch:

  .. code-block:: python

     class Example(Package):

         git = "https://github.com/example-project/example.git"

         version("develop")

  This download method is untrusted, and is not recommended. Aside from HTTPS,
  there is no way to verify that the repository has not been compromised, and
  the commit you get when you install the package likely won't be the same
  commit that was used when the package was first written. Additionally, the
  default branch may change. It is best to at least specify a branch name.

Branches
  To fetch a particular branch, use the ``branch`` parameter:

  .. code-block:: python

     version("experimental", branch="experimental")

>   This download method is untrusted, and is not recommended. Branches are

Reviewer 3: This seems like an overly harsh recommendation against a very common use case. What about

```suggestion
  This download method is untrusted, and is not recommended for production installations. Branches are
```

  moving targets, so the commit you get when you install the package likely
  won't be the same commit that was used when the package was first written.

Tags
  To fetch from a particular tag, use ``tag`` instead:

  .. code-block:: python

     version("1.0.1", tag="v1.0.1")

  This download method is untrusted, and is not recommended. Although tags
  are generally more stable than branches, Git allows tags to be moved.
  Many developers use tags to denote rolling releases, and may move the

>   tag when a bug is patched.

Reviewer 3: ```suggestion
  tag when a bug is patched. Instead, it is recommended to combine the `tag` and `commit`
  options (see below).
```

Commits
  Finally, to fetch a particular commit, use ``commit``:

  .. code-block:: python

>      version("2014-10-08", commit="9d38cd4e2c94c3cea97d0e2924814acc")

Reviewer 3: ```suggestion
     version("2014-10-08", commit="9d38cd4e2c94c3cea97d0e2924814acc")
     version("1.0.4", tag="v1.0.4", commit="8c4abe172c94c7c2aa7bd1292ad01bd7")
```

        # paths can be directories or functions, all subdirectories and files are included
        paths = ["doe", "rae", "me/file.cpp"]
        if package.spec.version >  Version("1.2.0"):
            paths.extend(["fae"])
        return paths

    class MyPackage(package):
        # can also be a package attribute that will be used if not specified in versions
        git_sparse_paths = ["doe", "rae"]

        # use the package attribute
        version("1.0.0")
        version("1.1.0")
        # use the function
        version("1.1.5", git_sparse_paths=sparse_path_func)
        version("1.2.0", git_sparse_paths=sparse_path_func)
        version("1.2.5", git_sparse_paths=sparse_path_func)
        version("1.1.5", git_sparse_paths=sparse_path_func)

.. _github-fetch:

""""""
GitHub
""""""

If a project is hosted on GitHub, *any* valid Git branch, tag, or hash
may be downloaded as a tarball.  This is accomplished simply by
constructing an appropriate URL.  Spack can checksum any package
downloaded this way, thereby producing a trusted download.  For
example, the following downloads a particular hash, and then applies a
checksum.

.. code-block:: python

>        version("1.9.5.1.1", md5="d035e4bc704d136db79b43ab371b27d2",

Reviewer 3: Should use a sha256 or other supported checksum method

    blis threads=openmp,pthreads

    Concretized
    --------------------------------
    ==> Error: multiple values are not allowed for variant "threads"

Another useful note is that *Python's* ``None`` *is not allowed as a default value*
and therefore it should not be used to denote that no feature was selected.
Users should instead select another value, like ``"none"``, and handle it explicitly
within the package recipe if need be:

  .. code-block:: python

      if self.spec.variants["threads"].value == "none":
         options.append("--no-threads")

In cases where multiple values can be selected at the same time ``multi`` should
be set to ``True``:

  .. code-block:: python

    class Gcc(AutotoolsPackage):
        ...
        variant(
            "languages", default="c,c++,fortran",
            values=("ada", "brig", "c", "c++", "fortran",
                    "go", "java", "jit", "lto", "objc", "obj-c++"),
            multi=True,
            description="Compilers and runtime libraries to build"
        )

Within a package recipe a multi-valued variant is tested using a ``key=value`` syntax:

  .. code-block:: python

    if spec.satisfies("languages=jit"):

>         options.append("--enable-host-shared")

Reviewer 3: ```suggestion
    if spec.satisfies("languages=jit"):
        options.append("--enable-jit")
```
I think this is clearer to the reader than the real example

that allows any combination of the specified values, and also allows the
user to specify ``"none"`` (as a string) to choose none of them.
The objects returned by these functions can be modified at will by chaining
method calls to change the default value, customize the error message or
other similar operations:

  .. code-block:: python

    class Mvapich2(AutotoolsPackage):
        ...
        variant(
            "process_managers",
            description="List of the process managers to activate",
            values=disjoint_sets(
                ("auto",), ("slurm",), ("hydra", "gforker", "remshell")
            ).prohibit_empty_set().with_error(
                "'slurm' or 'auto' cannot be activated along with "
                "other process managers"
            ).with_default("auto").with_non_feature_values("auto"),
        )

"""""""""""""""""""""""""""
Conditional Possible Values
"""""""""""""""""""""""""""

There are cases where a variant may take multiple values, and the list of allowed values
expands over time. Consider, for instance, the C++ standard with which we might compile
Boost, which can take one of multiple possible values with the latest standards

> only available from a certain version on.

Reviewer 3: ```suggestion
only available for more recent versions.
```

           # C++20/2a is not support by Boost < 1.73.0
           conditional("2a", "2b", when="@1.73.0:")
       ),
       multi=False,
       description="Use the specified C++ standard when building.",
   )

The snippet above allows ``98``, ``11`` and ``14`` as unconditional possible values for the
``cxxstd`` variant, while ``17`` requires a version greater or equal to ``1.63.0``
and both ``2a`` and ``2b`` require a version greater or equal to ``1.73.0``.

^^^^^^^^^^^^^^^^^^^^
Conditional Variants
^^^^^^^^^^^^^^^^^^^^

The variant directive accepts a ``when`` clause. The variant will only
be present on specs that otherwise satisfy the spec listed as the
``when`` clause. For example, the following class has a variant
``bar`` when it is at version 2.0 or higher.

.. code-block:: python

   class Foo(Package):
       ...
       variant("bar", default=False, when="@2.0:", description="help message")

The ``when`` clause follows the same syntax and accepts the same
values as the ``when`` argument of
:py:func:`spack.directives.depends_on`.

^^^^^^^^^^^^^^^
Sticky Variants
^^^^^^^^^^^^^^^

The variant directive can be marked as ``sticky`` by setting to ``True`` the

> corresponding argument:

Reviewer 3: ```suggestion
The variant directive can be marked as ``sticky`` by setting  the
corresponding argument to ``True``:
```

       ...

       depends_on("libelf")

       def install(self, spec, prefix):
           ...

^^^^^^^^^^^^^^^^
``depends_on()``
^^^^^^^^^^^^^^^^

The highlighted ``depends_on("libelf")`` call tells Spack that it
needs to build and install the ``libelf`` package before it builds
``libdwarf``.  This means that in your ``install()`` method, you are
guaranteed that ``libelf`` has been built and installed successfully,
so you can rely on it for your libdwarf build.

^^^^^^^^^^^^^^^^
Dependency specs
^^^^^^^^^^^^^^^^

``depends_on`` doesn't just take the name of another package. It can
take a full spec as well. This means that you can restrict the versions or
other configuration options of ``libelf`` that ``libdwarf`` will build
with. For example, suppose that in the ``libdwarf`` package you write:

.. code-block:: python

   depends_on("libelf@0.8")

Now ``libdwarf`` will require ``libelf`` in the range ``0.8``, which
includes patch versions ``0.8.1``, ``0.8.2``, etc. Apart from version
restrictions, you can also specify variants if this package requires
optional features of the dependency.

.. code-block:: python

   depends_on("libelf@0.8 +parser +pic")

> Both users *and* package authors can use the same spec syntax to refer

Reviewer 3: ```suggestion
Both users *and* package authors use the same spec syntax to refer
```

MPI has several different implementations (e.g., `MPICH <http://www.mpich.org>`_, `OpenMPI <http://www.open-mpi.org>`_, and `MVAPICH <http://mvapich.cse.ohio-state.edu>`_) and scientific applications can be built with any one of them.
Many package managers handle interfaces like this by requiring many variations of the package recipe for each implementation of MPI, e.g., ``foo``, ``foo-mvapich``, ``foo-mpich``.
In Spack every package is defined in a single ``package.py`` file, and avoids the combinatorial explosion through *virtual dependencies*.

^^^^^^^^^^^^
``provides``
^^^^^^^^^^^^

In Spack, ``mpi`` is handled as a *virtual package*.
A package like ``mpileaks`` can depend on the virtual ``mpi`` just like any other package, by supplying a ``depends_on`` call in the package definition.
For example:

.. code-block:: python
   :linenos:
   :emphasize-lines: 7

   class Mpileaks(Package):
       homepage = "https://github.com/hpc/mpileaks"
       url = "https://github.com/hpc/mpileaks/releases/download/v1.0/mpileaks-1.0.tar.gz"

       version("1.0", md5="8838c574b39202a57d7c2d68692718aa")

       depends_on("mpi")
       depends_on("adept-utils")
       depends_on("callpath")

> Here, ``callpath`` and ``adept-utils`` are concrete packages, but there is no actual package file for ``mpi``, so we say it is a *virtual* package.

Reviewer 3: ```suggestion
Here, ``callpath`` and ``adept-utils`` are concrete packages, but there is no actual package for ``mpi``, so we say it is a *virtual* package.
```
I would be careful not to say there is no package *file*, as there can be a package.py for a virtual (`mpi` has one to define tests shared across all implementations).

.. code-block:: console

   $ spack spec netlib-scalapack  ^[virtuals=lapack] openblas ^[virtuals=blas] atlas
   ==> Error: concretization failed for the following reasons:

      1. Package 'openblas' needs to provide both 'lapack' and 'blas' together, but provides only 'lapack'

^^^^^^^^^^^^^^^^^^^^
Versioned Interfaces
^^^^^^^^^^^^^^^^^^^^

Just as you can pass a spec to ``depends_on``, so can you pass a spec to ``provides`` to add constraints.
This allows Spack to support the notion of *versioned interfaces*.
The MPI standard has gone through many revisions, each with new functions added, and each revision of the standard has a version number.
Some packages may require a recent implementation that supports MPI-3 functions, but some MPI versions may only provide up to MPI-2.
Others may need MPI 2.1 or higher.  
You can indicate this by adding a version constraint to the spec passed to ``provides``:

.. code-block:: python

   provides("mpi@:2")

Suppose that the above ``provides`` call is in the ``mpich2`` package.
This says that ``mpich2`` provides MPI support *up to* version 2, but if a package ``depends_on("mpi@3")``, then Spack will *not* build that package with ``mpich2``.

> 

Reviewer 3: ```suggestion

Currently, versions and names are the only spec components supported for virtual packages.

```

.. _language-dependencies:

----------------------------------
Language and compiler dependencies
----------------------------------

Whenever you use ``spack create`` to create a new package, Spack scans the package's source code and heuristically adds *language dependencies*, which look like this:

.. code-block:: python

   depends_on("c", type="build")
   depends_on("cxx", type="build")
   depends_on("fortran", type="build")

The languages ``c``, ``cxx`` and ``fortran`` are **virtuals provided by compiler packages**, such as ``gcc``, ``llvm``, or ``intel-oneapi-compilers``.

When you concretize a package that depends on ``c``, Spack will select a compiler for it that provides the ``c`` virtual package.

Typically one compiler will be used to provide all languages, but Spack is allowed to create a mixed toolchain.
For example, the ``c`` compiler could be ``clang`` from the ``llvm`` package, whereas the ``fortran`` compiler could be ``gfortran`` from the ``gcc``.
This means that language dependencies translate to one or more compiler packages as build dependencies.

> 

Reviewer 3: Do we have a more detailed section on compiler dependencies we can link to here?

Author: I don't think so

|                                                                            | from sourceforge.org             |
+----------------------------------------------------------------------------+----------------------------------+
| :class:`~spack_repo.builtin.build_systems.sourceware.SourcewarePackage`    | A helper class for packages      |
|                                                                            | from sourceware.org              |
+----------------------------------------------------------------------------+----------------------------------+
| :class:`~spack_repo.builtin.build_systems.xorg.XorgPackage`                | A helper class for x.org         |
|                                                                            | packages                         |
+----------------------------------------------------------------------------+----------------------------------+

These mixins should be used as additional base classes for your package, in addition to the base class that you would normally use (e.g. ``MakefilePackage``, ``AutotoolsPackage``, etc.):

> 

Reviewer 3: It's probably worth clarifying the difference between `PythonPackage` and `PythonExtension` somewhere.

I'm assuming part 2 covers the build system classes in more detail? We should probably have a reference here to that section.

   # baz_home implementation provides the subdirectory inside the foo prefix
   >>> spec["baz"].home
   "/opt/spack/linux-fedora35-haswell/gcc-11.3.1/foo-1.0-ca3rczp5omy7dfzoqw4p7oc2yh3u7lt6/baz"

   # baz headers in the baz subdirectory of the foo prefix
   >>> spec["baz"].headers
   HeaderList([
       "/opt/spack/linux-fedora35-haswell/gcc-11.3.1/foo-1.0-ca3rczp5omy7dfzoqw4p7oc2yh3u7lt6/baz/include/baz/baz.h"
   ])

   # baz include directories in the baz subdirectory of the foo prefix
   >>> spec["baz"].headers.directories
   [
       "/opt/spack/linux-fedora35-haswell/gcc-11.3.1/foo-1.0-ca3rczp5omy7dfzoqw4p7oc2yh3u7lt6/baz/include"
   ]

   # baz libraries in the baz subdirectory of the foo prefix
   >>> spec["baz"].libs
   LibraryList([
       "/opt/spack/linux-fedora35-haswell/gcc-11.3.1/foo-1.0-ca3rczp5omy7dfzoqw4p7oc2yh3u7lt6/baz/lib/libFooBaz.so"
   ])

   # baz library directories in the baz subdirectory of the foo prefix
   >>> spec["baz"].libs.directories
   [
       "/opt/spack/linux-fedora35-haswell/gcc-11.3.1/foo-1.0-ca3rczp5omy7dfzoqw4p7oc2yh3u7lt6/baz/lib"
   ]

.. _make-package-findable:

----------------------------------------------------------

> Making a package discoverable with ``spack external find``

Reviewer 3: I think this whole section is missing a discussion of detection from `cls.libraries` instead of `cls.executables`.

Author: I would prefer a separate PR for that
